File: JavaBlockChainExample/src/main/java/com/dsc/blockchain/Block.java
/**
 * Block.java
 * 
 * JavaBlockChainExample Project is a blockchain example 
 * project to test the blockchain concept in Java.
 * 
 * Created by William Williams on 4/07/18.
 * Copyright © 2018 DSC. All rights reserved.
 */

package com.dsc.blockchain;

import java.lang.reflect.Array;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.Date;

public class Block 
{
	private String transactionPrevHash;
	
	private Integer transactionIndex;
	
	private Long powNonce;
	
	private String transactionData;
	
	private String transactionHash;
	
	private Date transactionTime;
	
	/**
	 * Constructor Method
	 * 
	 * @param transactionIndexIn
	 * @param transactionDataIn
	 */
	public Block(Integer transactionIndexIn, String transactionDataIn)
	{
		this.transactionIndex = transactionIndexIn;
		
		this.transactionData = transactionDataIn;
		
	    this.powNonce = -1L;
	    
	    this.transactionTime = new Date(System.currentTimeMillis());
	}
	
	/**
	 * setTransactionPrevHash Method
	 * 
	 * @param transactionPrevHash
	 */
	public void setTransactionPrevHash(String transactionPrevHash) 
	{
		this.transactionPrevHash = transactionPrevHash;
	}

	/**
	 * getTransactionHash Method
	 * 
	 * @return
	 */
	public String getTransactionHash() 
	{
		return transactionHash;
	}

	/**
	 * generateValidHash Method
	 * 
	 * "Proof of work" algorithm to generate a hash that start with difficulty number of zeros
	 * 
	 * @param difficulty
	 */
	public void generateValidHash(int difficulty)
	{
		Instant validationStartTime = Instant.now();
		
		String[] cString = (String[]) Array.newInstance(String.class, difficulty);
		
		// Create POW string array of zeros (difficulty) long
		for (int i=0; i < difficulty; i++)
		{
			cString[i] = "0";
		}
		
		do
		{
			transactionHash = calculateHash();
		}
		while (!transactionHash.substring(0, difficulty).equals(cString));
		
		Instant validationEndTime = Instant.now();
		
		Duration validationDuration = Duration.between(validationStartTime, validationEndTime);
        
		System.out.print("Block Minded ["+transactionHash+"] StartDT["+ DateTimeFormatter.ISO_INSTANT.format(validationStartTime) +"] EndTime["+ DateTimeFormatter.ISO_INSTANT.format(validationEndTime) +"] Total Duration["+validationDuration.toString()+"]\n\n");
	}
	
	/**
	 * CalculateHash Method
	 * 
	 * @return
	 */
	private String calculateHash()
	{
	    String hashString = null;
	    
	    String retHashString = null;

	    hashString = this.transactionIndex.toString() + this.transactionTime +this.transactionData +this.powNonce + this.transactionPrevHash;
	    
	    MessageDigest mbDigest;
	    
		try 
		{
			mbDigest = MessageDigest.getInstance("SHA-256");
			
			mbDigest.update(hashString.getBytes());
			
			byte[] tempHash = mbDigest.digest();
			
		    StringBuffer hashBuffer = new StringBuffer();

		    for(int i=0;i<tempHash.length;i++)
		    {
		    	hashBuffer.append(Integer.toString((tempHash[i] & 0xff) + 0x100, 16).substring(1));//tempHash[i]);
		    }
			
			retHashString = hashBuffer.toString();
		} 
		catch (NoSuchAlgorithmException e) 
		{
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	    
	    return retHashString;
	}
}

File: CppBlockChainExample/Block.cpp
//
//  Block.cpp
//  Blockchain
//
//  Created by William Williams on 3/22/18.
//  Copyright © 2018 DSC. All rights reserved.
//

#include <stdio.h>
#include <time.h>
#include <sstream>
#include "Block.h"
#include "./sha/sha256.h"

/*
 * Constructor
 */
Block::Block(uint32_t transactionIndexIn, const string &transactionDataIn) : transactionIndex(transactionIndexIn), transactionData(transactionDataIn)
{
    cout << "New Block with Index[ " << transactionIndexIn << " Data[" << transactionDataIn << "]" << endl;
    
    powNonce = -1;
    
    transactionTime = time(nullptr);
    
    transactionHash = _CalculateHash();
}

/*
 * Getters & Setters
 */
string Block::GetTransactionHash()
{
    return transactionHash;
}

void Block::SetTransactionPrevHash(const string &transactionPrevHashIn)
{
    transactionPrevHash = transactionPrevHashIn;
}



/*
 * "Proof of work" algorithm to generate a hash that start with nDifficulty number of zeros
 */
void Block::GenerateValidHash(uint32_t nDifficulty)
{
    time_t validationStartTime;
    
    time_t validationEndTime;
    
    time(&validationStartTime);
    
    char buffer [80];
    
    struct tm *tmp ;
    
    string strStartTime, strEndTime, strDiffTime;
    
    char cstr[nDifficulty + 1];
    
    for (uint32_t i=0; i< nDifficulty; ++i)
    {
        cstr[i] ='0';
    }
    
    cstr[nDifficulty] = '\0';
    
    string str(cstr);
    
    do
    {
        powNonce++;
        
        transactionHash = _CalculateHash();
    }
    while (transactionHash.substr(0, nDifficulty) != str);
    
    time(&validationEndTime);
    
    double diff_t = difftime(validationEndTime, validationStartTime);
    
    // Pretty Print StartTime
    tmp = localtime(&validationStartTime);
    
    strftime(buffer,80,"Now it's %I:%M%p.",tmp);
    
    puts (buffer);
    
    strStartTime = buffer;

    
    // Pretty Print EndTime
    tmp = localtime(&validationEndTime);
    
    strftime(buffer,80,"Now it's %I:%M%:%S%p.",tmp);
    
    puts (buffer);
    
    strEndTime = buffer;

    
    cout << "Block mined: " << transactionHash << " Start Time: ["<<strStartTime<<" ] EndTime: ["<<strEndTime<<"] Total Time to mine ="<<diff_t <<endl;
}

/*
 * Calculate hash function
 */
inline string Block::_CalculateHash() const
{
    stringstream ss;

    ss << transactionIndex << transactionTime << transactionData << powNonce << transactionPrevHash;
    
    return sha256(ss.str());
}
